### DISCUSSION TITLE
FTPS接続時にCRLチェックはFFFTPの制御か？

### DISCUSSION THREAD

**Discussion by Hirose-Wakaba** [ffftp/ffftp#394] [Q&A]
大概のブラウザでは設定でCRL/OCSPのチェックを有効・無効化は可能ではあるが、FFFTPでは同様な設定は出来ないのか？
インターネットへのオフライン環境下ではアクセス時に問題が生じる(CRL/OCSPを読み込めないからだと思われる)。
FTPSの接続は出来ているようにみえるが、以下のログは正常に暗号化しているとみて問題ないか？
また、CRL/OCSPのチェックが出来ない場合も証明書自体が正常だと認識しない模様だが、それは仕様か？

<pre>
FTP over Explicit SSL/TLS (FTPES)を使用します.
ホスト ftp.example.local (192.168.100.77:21) に接続しています.
接続しました.
220 Microsoft FTP Service
>AUTH TLS
234 AUTH command ok. Expecting TLS Negotiation.
証明書の検証エラー: 0x80092012: 失効の関数は証明書の失効を確認できませんでした。 
>PBSZ 0
200 PBSZ command successful.
>PROT P
200 PROT command successful.
>USER FTPUser
331 Password required
>PASS [xxxxxx]
230 User logged in.
>FEAT
211-Extended features supported:
 LANG EN*
 UTF8
 AUTH TLS;TLS-C;SSL;TLS-P;
 PBSZ
 PROT C;P;
 CCC
 HOST
 SIZE
 MDTM
 REST STREAM
211 END
>OPTS UTF8 ON
200 OPTS UTF8 command successful - UTF8 encoding now ON.
>XPWD
257 "/" is current directory.
>TYPE A
200 Type set to A.
>PASV
227 Entering Passive Mode (192.168.100.77,208,97).
ホスト 192.168.100.77 (192.168.100.77:53345) に接続しています.
接続しました.
>LIST
125 Data connection already open; Transfer starting.
証明書の検証エラー: 0x80092012: 失効の関数は証明書の失効を確認できませんでした。 
226 Transfer complete.
ファイル一覧の取得は正常終了しました. (0 Bytes)
</pre>

**Comment by sayurin** [comment]
現状では仕様です。

https://github.com/ffftp/ffftp/blob/731e2259e3709d46a1a204eb949975edb8646aa3/socket.cpp#L172

ここで[`CERT_CHAIN_REVOCATION_CHECK_CHAIN_EXCLUDE_ROOT`](https://learn.microsoft.com/ja-jp/windows/win32/api/wincrypt/nf-wincrypt-certgetcertificatechain#:~:text=CERT_CHAIN_REVOCATION_CHECK_CHAIN_EXCLUDE_ROOT)を指定しており、

> 失効チェックは、ルート証明書を除くすべてのチェーン内のすべての証明書で行われます。

という動作をします。失効チェックできるべきですし、できないようなオレオレ証明書なら別段セキュリティに寄与しないので、平文通信すればいいかなという判断です。

----

> インターネットへのオフライン環境下ではアクセス時に問題が生じる(CRL/OCSPを読み込めないからだと思われる)。

この部分はよくわかりませんでした。オンラインなら問題なくオフラインだから問題が生じる、ということでしょうか？
CRLがきちんと運用されていないオレオレ証明書が原因でオンライン・オフラインは無関係かと考えていました。
失効チェックに対する解釈に誤りがあるのであれば、仕様を見直したいと思います。

**Reply by Hirose-Wakaba** [reply]
なるほど、現状では仕様なのですね…中々情報が無かったので、大変助かりました。ご返答ありがとうございます

証明書内にCRLのディストリビューションポイント…要するにCRL(失効ファイル一覧)を参照出来るURIが証明書内に記述する
のはご存じかと思いますが、そこにアクセスできないと結局エラーになります。
CRL込みの証明書でも前述のエラーとは異なりますが、**CRLを記載しても「失効サーバーがオフラインのため、失効の関数は**
**失効を確認できませんでした」** というエラーメッセージに変化します。
尚、CRL込みの証明書であればオンライン環境下に持ってくると正常でした。

失効チェック自体を入れない理由は失効リストを持たせるサーバを別途用意しておく必要があります。OSのキャッシュ期間を
過ぎると再アクセスが発生しますので、中々ダウンさせられないシステムになります(あと、運用も面倒ｗ)。
ならばCRL/OCSPの項目は自己証明書としては不要だというスタンスから外している次第です。ブラウザやメーラはCRL/OCSP
の有無は厭わないようなので、同じ方向性だと有難いですね。

参考事例
https://ssl.sakura.ad.jp/column/ssl-failure/

CRLに関してはご検討いただけると助かりますが、現時点では仕様であることが判明出来たのが最大のポイントでしたので
この件はクローズで問題ありません。

ご教示頂き大変ありがとうござました。
